tracing/src/main/java/com/palantir/tracing/TagRecorder.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import java.util.Map;
+
+/**
+ * Interface which allows tracing spans to be provided with tag metadata. This interface
+ * is structured such that implementations can be created once and reused, given {@link S data}
+ * from which all tag values are extracted. Unlike other approaches, this doesn't require per-span
+ * lambdas and has minimal overhead in both the sampled and unsampled cases. This allows unsampled spans
+ * to result in zero allocation.
+ *
+ * @param <S> Stateful object type, an instance of which is passed through method calls with the {@link TagRecorder}.
+ */
+public interface TagRecorder<S> {
```

Comment:
Naming is a little funny here:

Tracing has had an (unused) concept of `map<string,string>` metadata for years, so I retained the metadata name to describe the map, however here I call this a TagRecorder rather than a MetadataRecorder. I'd appreciate a second opinion. Maybe I rename all the `metadata` map parameter names to `tags`?

I bobbled between a few suffixes before using `Recorder`. I think recorder is probably fine, but `Translator` could also make sense. What do you think?

Reply:
I think Recorder is good, maybe Translator vibes more with me, mostly because of this precedent: https://github.com/LMAX-Exchange/disruptor/blob/master/src/main/java/com/lmax/disruptor/EventTranslator.java#L27.

I would also maybe say TagAdapter -> TagSink

Reply:
I borrowed the translator idea from disruptor ;-) I'll rename to translator since I think it provides more meaning.

> I would also maybe say TagAdapter -> TagSink

Hmm, sink doesn't feel quite right because it isn't a destination, so much as a pass-through to `T target` which is the sink. Although that may not help readability....

I definitely need to add more javadoc in this interface.